doc/developer/platform/ux.md

```diff
@@ -285,31 +285,152 @@ The HTTPS server exposes two APIs:
+        "BEGIN;
+        INSERT INTO mytable VALUES(1);
+        SELECT * FROM mytable;
+        COMMIT;
+        INSERT INTO mytable VALUES(2);
+        SELCT 1/0;"
+    }'
+    ```
+
+    The response would contain all of the results from the first transaction,
+    and only contain an error for the remaining statements.
+
+    ```
+    [<BEGIN tag>, <INSERT tag>, <rows from SELECT>, <COMMIT ack>, <divison by 0 error>]
+    ```
+
+    This means that one cannot directly correlate the statement's index with the
+    response's indices to determine the statement's values, particularly in
+    instances where there are errors.
 
     In the future, we may expand the feature set of the HTTPS API to allow
     additional control over query execution and result fetching, a la the
     [Snowflake SQL REST API].
 
-    TODO: fully describe the output format.
+    ### Output format
+
+    The output format is a JSON object with one key, `results`, whose values is
+    an array of serialized `adapter::command::SimpleResult`, representing all of
+    the committed statements within each executed transaction. Some transactions
+    might not be executed because we aborted the request early due to
+    encountering an error.
+
+    The serialized `adapter::command::SimpleResult` values are representing in JSON as:
+
+    `SimpleResult` value | JSON value
+    ---------------------|------------
+    `Rows` | `{"rows": <2D array of JSON-ified results>, "col_names": <array of text>}`
+    `Error` | `{"error": <Error string from execution>}`
+    `Ok` | `{ "ok": <tag>}`
+
+    Each committed statement returns exactly one of these values; e.g. in the
+    case of "complex responses", such as `INSERT INTO...RETURNING`, the presence
+    of `Rows` implies `Ok`. In the future we can consider including the tagged
+    `Ok`.
+
+    ### Binding parameters
+
+    You can include parameters alongside the query by including on the request
+    body a key `"parameters"` whose value is an array of objects of the form:
+
+    ```json
+    {
+      "type": "<resolvable type name>",
+      "value": "<parseable text value>"
+    }
+    ```
+
+    You can then reference these parameters in the query as you would with PG,
+    i.e. using 1-indexed references to the parameters array, prefaced by a `$`. e.g.
 
-    TODO: design support for binding parameters.
+    ```json
+    {
+      "sql": "INSERT INTO foo VALUES ($1);",
+      "parameters": [{"type": "int", "value": "1"}]
+    }
+    ```
+
+    Note that all values must be strings; the API will not accept JSON numbers.
 
   * `ws://<ENDPOINT>/query?query=QUERY` streams the result of a single query
     over the WebSockets protocol.
 
     This API is particularly useful for listening to the results of `TAIL`
     queries from a browser.
 
-    TODO: fully describe the output format.
+    ### Output
+
+    A Websocket request outputs the following:
+      - `{"ok": <tag>}` once the command has completed successfully; the `tag`
```

Comment:
By what means do we provide access to those in the current `client.rs` code?

https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/client.rs#L427-L433

Reply:
Just noting that this is the big open question in terms of generating more meaningful `ok` messages.

Reply:
Hrm these currently get made in protocol.rs: https://github.com/MaterializeInc/materialize/blob/405f7e8bd9dddb67a8a2bf429205107dac482001/src/pgwire/src/protocol.rs#L1836. Even for things like SELECT we either have to do some annoying refactoring, or else reimplement these all here. So I don't currently think there's any reason to ever produce null, because it's not like we have the tag for the non-rows producing statements anyway.

Reply:
Matt and I chatted about this and the idea here is that we can just invent tags where necessary, i.e something like `execute_response.tag().unwrap_or(||match stmt...)`